cli/context/store/store.go

```diff
@@ -0,0 +1,329 @@
+	if err := os.MkdirAll(metaRoot, 0755); err != nil {
+		return nil, err
+	}
+	if err := os.MkdirAll(tlsRoot, 0700); err != nil {
+		return nil, err
+	}
+	_, err := os.Stat(configFile)
+
+	switch {
+	case os.IsNotExist(err):
+		//create default file
+		if err := ioutil.WriteFile(configFile, []byte("{}"), 0644); err != nil {
+			return nil, err
+		}
+	case err != nil:
+		return nil, err
+	}
+
+	configBytes, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return nil, err
+	}
+	var cfg config
+	if err := json.Unmarshal(configBytes, &cfg); err != nil {
+		return nil, err
+	}
+	return &store{
+		configFile:     configFile,
+		currentContext: cfg.CurrentContext,
+		meta: &metadataStore{
+			root: metaRoot,
+		},
+		tls: &tlsStore{
+			root: tlsRoot,
+		},
+	}, nil
+}
+
+type store struct {
+	configFile     string
+	currentContext string
+	meta           *metadataStore
+	tls            *tlsStore
+}
+
+func (s *store) GetCurrentContext() string {
+	return s.currentContext
+}
+
+func (s *store) SetCurrentContext(name string) error {
+	cfg := config{CurrentContext: name}
+	configBytes, err := json.Marshal(&cfg)
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(s.configFile, configBytes, 0644); err != nil {
+		return err
+	}
+	s.currentContext = name
+	return nil
+}
+
+func (s *store) ListContexts() (map[string]ContextMetadata, error) {
+	return s.meta.list()
+}
+
+func (s *store) CreateOrUpdateContext(name string, meta ContextMetadata) error {
+	return s.meta.createOrUpdate(name, meta)
+}
+
+func (s *store) RemoveContext(name string) error {
+	if err := s.meta.remove(name); err != nil {
+		return err
+	}
+	return s.tls.removeAllContextData(name)
+}
+
+func (s *store) GetContextMetadata(name string) (ContextMetadata, error) {
+	return s.meta.get(name)
+}
+
+func (s *store) ResetContextTLSMaterial(name string, data *ContextTLSData) error {
+	if err := s.tls.removeAllContextData(name); err != nil {
+		return err
+	}
+	if data == nil {
+		return nil
+	}
+	for ep, files := range data.Endpoints {
+		for fileName, data := range files.Files {
+			if err := s.tls.createOrUpdate(name, ep, fileName, data); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *store) ResetContextEndpointTLSMaterial(contextName string, endpointName string, data *EndpointTLSData) error {
+	if err := s.tls.removeAllEndpointData(contextName, endpointName); err != nil {
+		return err
+	}
+	if data == nil {
+		return nil
+	}
+	for fileName, data := range data.Files {
+		if err := s.tls.createOrUpdate(contextName, endpointName, fileName, data); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *store) ListContextTLSFiles(name string) (map[string]EndpointFiles, error) {
+	return s.tls.listContextData(name)
+}
+
+func (s *store) GetContextTLSData(contextName, endpointName, fileName string) ([]byte, error) {
+	return s.tls.getData(contextName, endpointName, fileName)
+}
+
+// Export exports an existing namespace into a stream
```

Comment:
IMO, we should consider it opaque as in "undocumented", but backward compatible. In the future, if we choose to have a tlsStore that interacts with the Windows cert store, or MacOS keychain, we'll keep the same tar based format for import/export, but the way tlsStore will persist them at rest will be quite different.

Reply:
Actually the tarball has not the exact same layout as the context store implementation: it does not have the contextName subdir in meta/ and tls/ roots.

Reply:
What do you think of prepending some sort of non-tar header so that tools like `file(1)` or file managers etc do not think it is a tarball and people aren't tempted to pick it apart? As a bonus you could include a version field there to give some wriggle room if this ever needs to change format (you could also use a custom PAXHeader to achieve that though).

OTOH for debugging and things being able to `tar xf` is probably quite useful.

Reply:
I am not sure yet about that. We might want at some point to document the format, so that cloud provider can provide managed Docker Hosting and emit Docker Contexts in this format instead of scripts based bundles. @thaJeztah wdyt ?